scripts/cache_vision_features.py

```diff
@@ -0,0 +1,47 @@
+import os
+import glob
+import argparse
+from tqdm import tqdm
+
+import torch
+
+from allennlp.common.file_utils import TensorCache, cached_path
+from allennlp.data import DetectronImageLoader
+from allennlp.modules.vision import ResnetBackbone, FasterRcnnRegionDetector
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--image_dir", type=str)
+    parser.add_argument("--cache_path", type=str)
+    parser.add_argument("--use-cuda", action="store_true", help="use GPU if one is available")
+    parser.add_argument("--batch-size", type=int, default=32)
+    args = parser.parse_args()
+    return args
+
+if __name__ == "__main__":
+    args = parse_args()
+    os.makedirs(args.cache_path, exist_ok=True)
+    features_cache = TensorCache(os.path.join(args.cache_path, "features"))
+    coordinates_cache = TensorCache(os.path.join(args.cache_path, "coordinates"))
+    image_paths = list(glob.iglob(os.path.join(args.image_dir, "**", "*.png"), recursive=True))
+    image_keys = [os.path.basename(filename) for filename in image_paths]
+
+    image_loader = DetectronImageLoader()
+    image_featurizer = ResnetBackbone()
+    region_detector = FasterRcnnRegionDetector()
+    if torch.cuda.is_available() and args.use_cuda:
+        backbone.cuda()
+        region_detector.cuda()
+
+    images, sizes = image_loader(image_paths)
```

Comment:
This loads all the images at the same time. Does that fit?

Also, how much faster is it to run this than running just 2 images at a time (which is what the NLVR reader does right now)?

Reply:
Actually loading all images at once does give an error when using a GPU, but it's unclear to me why the image loader interacts with the GPU at all...

I found that caching the NLVR2 dev set takes 39 minutes with a batch size of 16 (batch size of 32 didn't fit on 24 GB GPU). How does this compare with what you found for doing it 2 at a time? [I changed it so that images are loaded in batches to make this work, but still wondering about why loading all at once causes a problem]

Reply:
Doing the training set took about 16 hours I think.

Reply:
Ok, extrapolating the dev set time from this gives 8103*16/103171 \approx 1.25 hours, so I think there is a speed up here.